src/density_methods.jl

```diff
@@ -134,50 +133,63 @@ function atomic_spin_density(basis::PlaneWaveBasis{T}, method::AtomicDensity,
         magmom[3] / n_elec_valence(atom)
     end::AbstractVector{T}  # Needed to ensure type stability in final guess density
 
-    form_factors = atomic_density_form_factors(basis, method)
-    atomic_density_superposition(basis, form_factors; coefficients)
+    atomic_density_superposition(basis, method; coefficients)
 end
 
 # Perform an atomic density superposition. The density is constructed in reciprocal space
 # using the provided atomic form-factors and coefficients and applying an inverse Fourier
 # transform to yield the real-space density.
 function atomic_density_superposition(basis::PlaneWaveBasis{T},
-                                      form_factors::IdDict{Tuple{Int,T},T};
+                                      method::AtomicDensity;
                                       coefficients=ones(T, length(basis.model.atoms))
                                       ) where {T}
-    model = basis.model
-    G_cart = to_cpu(G_vectors_cart(basis))
-    ρ_cpu = map(enumerate(to_cpu(G_vectors(basis)))) do (iG, G)
-        Gnorm = norm(G_cart[iG])
-        ρ_iG = sum(enumerate(model.atom_groups); init=zero(complex(T))) do (igroup, group)
-            sum(group) do iatom
-                structure_factor = cis2pi(-dot(G, model.positions[iatom]))
-                coefficients[iatom] * form_factors[(igroup, Gnorm)] * structure_factor
-            end
+    form_factors, iG2ifnorm = atomic_density_form_factors(basis, method)
+
+    # Pre-allocation of large arrays for GPU efficiency
+    Gs = G_vectors(basis)
+    ρ = to_device(basis.architecture, zeros(Complex{T}, length(Gs)))
+    ρ_tmp = similar(ρ)
+    indices = to_device(basis.architecture, collect(1:length(Gs)))
+
+    for (igroup, group) in enumerate(basis.model.atom_groups)
+        for iatom in group
+            r = basis.model.positions[iatom]
+            ff_group = @view form_factors[:, igroup]
+            map!(iG -> cis2pi(-dot(Gs[iG], r)) * ff_group[iG2ifnorm[iG]], ρ_tmp, indices)
+            ρ .+= ρ_tmp .* (coefficients[iatom] / sqrt(basis.model.unit_cell_volume))
         end
-        ρ_iG / sqrt(model.unit_cell_volume)
     end
-    ρ = to_device(basis.architecture, ρ_cpu)
+
     enforce_real!(ρ, basis)  # Symmetrize Fourier coeffs to have real iFFT
-    irfft(basis, ρ)
+    irfft(basis, reshape(ρ, basis.fft_size))
 end
 
+"""
+Returns the form factors at unique values of |G + q| (in Cartesian coordinates).
+Additionally, returns a mapping from any G index to the corresponding entry in the form_factors array.
+"""
 function atomic_density_form_factors(basis::PlaneWaveBasis{T},
```

Comment:
Is there some point in generalising this with the function you adapted in  #1056. I mean it is pretty much the same code or am I missing something ?

Reply:
Indeed, it is very similar. Only difference is the actual method for the calculation of the form factors (and the optional + q for atomic_local). I guess a single function could be used, with the actual form factor function name passed as (keyword) argument.

Given the order of file inclusion on `DFTK.jl`, it would probably need to be defined in `local.jl` or some other file up the tree.

Reply:
Or a separate file atomic_form_factors.jl or similar ... there is some point in using this for computing projectors for DOS calculations as well if I don't misremember.

Reply:
There might also be some application for the `AtomicNonlocal` term, which currently computes form factors for all G vectors, and not only for unique norms. I'd suggest to tackle this in a separate PR